nano-sync/src/circuits/mnt4/macro_block.rs

```diff
@@ -0,0 +1,425 @@
+                commitment.push(UInt8::from_bits_le(&pedersen_bits[i * 8..(i + 1) * 8]));
+            }
+
+            prepare_agg_pk_chunks_commitments.push(commitment);
+        }
+
+        // Calculating the commit aggregate public key. All the chunks come with the generator added,
+        // so we need to subtract it in order to get the correct aggregate public key. This is necessary
+        // because we could have a chunk of public keys with no signers, thus resulting in it being
+        // zero.
+        next_cost_analysis!(cs, cost, || { "Calculate commit agg pk" });
+
+        let mut commit_agg_pk = sum_generator_g2_var.clone();
+
+        for i in 0..self.commit_agg_pk_chunks.len() {
+            commit_agg_pk = commit_agg_pk.add(
+                cs.ns(|| format!("add next key, commit {}", i)),
+                &commit_agg_pk_chunks_var[i],
+            )?;
+
+            commit_agg_pk = commit_agg_pk.sub(
+                cs.ns(|| format!("subtract generator, commit {}", i)),
+                &sum_generator_g2_var,
+            )?;
+        }
+
+        commit_agg_pk = commit_agg_pk.sub(
+            cs.ns(|| "subtract generator, commit"),
+            &sum_generator_g2_var,
+        )?;
+
+        // Calculating the commitments to each of the commit aggregate public keys chunks. These
+        // will be given as input to the PKTree SNARK circuit.
+        next_cost_analysis!(cs, cost, || {
+            "Calculate commit agg pk chunks commitments"
+        });
+
+        let mut commit_agg_pk_chunks_commitments = Vec::new();
+
+        for i in 0..commit_agg_pk_chunks_var.len() {
+            let chunk_bits = SerializeGadget::serialize_g2(
+                cs.ns(|| format!("serialize commit agg pk chunk {}", i)),
+                &commit_agg_pk_chunks_var[i],
+            )?;
+
+            let pedersen_commitment = PedersenCommitmentGadget::evaluate(
+                cs.ns(|| format!("pedersen commitment commit agg pk chunk {}", i)),
+                &chunk_bits,
+                &pedersen_generators_var,
+                &sum_generator_g1_var,
+            )?;
+
+            let pedersen_bits = SerializeGadget::serialize_g1(
+                cs.ns(|| format!("serialize pedersen commitment, commit chunk {}", i)),
+                &pedersen_commitment,
+            )?;
+
+            let pedersen_bits = reverse_inner_byte_order(&pedersen_bits[..]);
+
+            let mut commitment = Vec::new();
+
+            for i in 0..pedersen_bits.len() / 8 {
+                commitment.push(UInt8::from_bits_le(&pedersen_bits[i * 8..(i + 1) * 8]));
+            }
+
+            commit_agg_pk_chunks_commitments.push(commitment);
+        }
+
+        // Preparing the inputs for the SNARK proof verification. All the inputs need to be Vec<UInt8>,
+        // so they need to be converted to such.
+        next_cost_analysis!(cs, cost, || { "Prepare inputs for SNARK" });
+
+        let position = vec![UInt8::constant(0)];
```

Comment:
If the position for the root is always [0], why do we need to pass it?
We don't need a `Vec<Uint8>` at least.

Reply:
We pass it because I didn't want to hardcode it into pk_tree_0. Anyway, if I'm going to merge pk_tree 0/2/4 to avoid code duplication, then I need it to have the position as an input.